02_bubble/ailyntang/sort.go

```diff
@@ -0,0 +1,44 @@
+package main
+
+import (
+	"fmt"
+	"io"
+	"os"
+)
+
+var out io.Writer = os.Stdout
+
+func main() {
+	fmt.Fprintln(out, bubble([]int{3, 2, 1, 5}))
+}
+
+func bubble(s []int) []int {
+	if len(s) < 2 {
+		return s
+	}
+
+	beforeSort := make([]int, len(s))
+	copy(beforeSort, s)
+	afterSort := make([]int, len(s))
+	copy(afterSort, s)
+
+	sortAdjacentPairs(afterSort)
+
+	for i, num := range afterSort {
+		if num != beforeSort[i] {
```

Comment:
I don't think this for loop is doing what you expect. It checks at indexes in the sorted slice and then calls the sorting function again if numbers don't match after sorting. But the fact that there has been a swap doesn't actually mean another sort is required. I would suggest addressing my other comments about test coverage etc and then you'll be in a better place to debug this logic and figure what it's doing wrong.

Reply:
Thanks for the call out. I realised I needed to reset the index of the for loop so that every element of the before and after slices are properly compared to each other.